### DISCUSSION TITLE
Building mpfs_icicle application examples fails with zephyr-sdk-0.15.0

### DISCUSSION THREAD

**Discussion by ChrRudel** [polarfire-soc/polarfire-soc-documentation#209] [RTOS for PolarFire SoC] [answered]
When trying to build the mpfs_icicle application example (synchronization, synchronization_blinky, smp/pi) the zephyr-sdk-0.15.0 does not complete the build but errors out with an incompatible ABI error. With zephyr-sdk-0.13.2 the application examples build without issues.

For example when trying to build the pi example with zephyr-sdk-0.15.0:
west build -p -b mpfs_icicle mpfs-applications/smp/pi -G "Eclipse CDT4 - Unix Makefiles"

Build fails at 93% Progress with the following errors:
[ 93%] Linking C executable zephyr_prebuilt.elf
/root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_clzdi2.o): ABI is incompatible with that of the selected emulation:
  target emulation `elf32-littleriscv' does not match `elf64-littleriscv'
/root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: failed to merge target specific data of file /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_clzdi2.o)
/root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_ctzdi2.o): ABI is incompatible with that of the selected emulation:
  target emulation `elf32-littleriscv' does not match `elf64-littleriscv'
/root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: failed to merge target specific data of file /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_ctzdi2.o)
/root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_clz.o): ABI is incompatible with that of the selected emulation:
  target emulation `elf32-littleriscv' does not match `elf64-littleriscv'
/root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: failed to merge target specific data of file /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_clz.o)
collect2: error: ld returned 1 exit status
make[2]: *** [zephyr/CMakeFiles/zephyr_prebuilt.dir/build.make:112: zephyr/zephyr_prebuilt.elf] Error 1
make[1]: *** [CMakeFiles/Makefile2:2524: zephyr/CMakeFiles/zephyr_prebuilt.dir/all] Error 2
make: *** [Makefile:91: all] Error 2

**Comment by con-pax** [comment]
Hi @ChrRudel, 
Thanks for pointing this out! We are currently taking the backports to v2.7-branch of Zephyr and applying it. Will update after testing.

**Comment by TommyMurphyTM1234** [comment]
>  Linking C executable zephyr_prebuilt.elf /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: /root/zephyr-sdk-0.15.0/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/12.1.0/libgcc.a(_clzdi2.o): ABI is incompatible with that of the selected emulation: target emulation `elf32-littleriscv' does not match `elf64-littleriscv'

These type of errors strongly suggest that the code is being compiled for an rv32 architecture but at link time the libraries for that architecture are not available. In that case the linker will fall back to the toolchain default architecture libraries (usually rv64gc unless overridden at toolchain configure time) and the rv32/rv64 mismatch arises. This can happen if the toolchain was not configured with `--enable-multilib` or the default set of multilibs (`gcc/gcc/config/riscv/t-elf-multilib`) does not specify the required rv32 architecture/ABI. However, the lack of a complete build log and details of the toolchain provenance and configuration means that some of this is informed speculation.

**Reply by AntonKrug** [reply]
Something similar (but in context of SoftConsole) is mentioned here as well:

https://mi-v-ecosystem.github.io/SoftConsole-Documentation/SoftConsole-v2022.2-RISC-V/troubleshooting/riscv.html#cc1-exe-error-march-your-architecture-invalid-isa-string

**Comment by con-pax** [comment, marked as answer]
Hi @ChrRudel,

The Zephyr kernel code base in the polarfire-soc-zephyr-applications is at v2.7, (which is an LTS version), plus our preliminary SMP support.
The latest version of Zephyr is v3.2.x.
from the Zephyr toolchain repo https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.15.0

Caveats

    Zephyr SDK 0.15.0 is only compatible with the Zephyr development branch (main) and the Zephyr release v3.2 and above. For more details, refer to the [Zephyr SDK Version Compatibility Matrix](https://github.com/zephyrproject-rtos/sdk-ng/wiki/Zephyr-SDK-Version-Compatibility-Matrix).
    RISC-V toolchain defaults to the ISA specification version 20191213, which requires the Zicsr and Zifencei extensions to be specified to the -march flag.

We are currently in the process of preparing our SMP support to be upstreamed to the Zephyr project. In the interim, we will push a branch with working examples and the latest version of the Zephyr code base to the polarfire-soc-zephyr-applications repository.

**Reply by con-pax** [reply]
A branch has been created with our latest SMP support targeting DDR and experimentally, Scratchpad. (currently in upstreaming process) [here](https://github.com/polarfire-soc/polarfire-soc-zephyr-applications/tree/zephyr-mpfs-smp_scratchpad)
Marking this issue as fixed. if it doesn't fix the issue add a comment and let us know.

**Reply by ChrRudel** [reply]
Hello @con-pax, 
I tried building the .polarfire-soc-zephyr-applications-zephyr-mpfs-smp_scratchpad application using this new branch and Zephyr SDK 0.15.0 and the build completes without errors. Thanks.
